content/en/docs/internals/testing.md

```diff
@@ -0,0 +1,432 @@
+documentation][services_networking]).
+
+Outside of a Kubernetes cluster, the connection information can be passed using
+the standard `KUBECONFIG` environment variable, which should point to a file
+just as `kubectl` and/or `oc` would expect.  Note that the current context will
+be used, so make sure it points to the correct cluster.
+
+{{< highlight bash >}}
+$ oc whoami
+error: Missing or incomplete configuration info.  Please point to an existing, complete config file:
+
+
+  1. Via the command-line flag --kubeconfig
+  2. Via the KUBECONFIG environment variable
+  3. In your home directory as ~/.kube/config
+
+To view or setup config directly use the 'config' command.
+$ export KUBECONFIG=…
+$ oc whoami --show-context
+bbguimaraes/api-build01-ci-devcluster-openshift-com:6443/bbarcaro
+{{< / highlight >}}
+
+Our `ci-operator` is finally ready to execute its test, but one last adjustment
+will make it a more pleasant experience.  In Kubernetes, isolation between
+workloads is achieved using [_namespaces_][namespaces].  Openshift wraps
+upstream namespaces into its own resource, the [_project_][project].  For most
+purposes, the distinction is irrelevant and both names are used interchangeably.
+`ci-operator` uses an ephemeral namespace to guarantee no interference between
+unrelated tests, which is referred to as the _test namespace_.  The process by
+which this namespace is set up will be discussed later; of relevance here is
+that this setup is inconvenient for manual testing, for various reasons.
+
+We can create our own test namespace in advance and simply ask `ci-operator` to
+use it instead of attempting to create one.  The build clusters are shared
+production infrastructure, so make sure to use a unique name that makes the
+owner and the purpose clear.
+
+{{< highlight bash >}}
+$ oc new-project bbguimaraes
+Now using project "bbguimaraes" on server "https://api.build01.ci.devcluster.openshift.com:6443".
+
+You can add applications to this project with the 'new-app' command. For example, try:
+
+    oc new-app rails-postgresql-example
+
+to build a new example application in Ruby. Or use kubectl to deploy a simple Kubernetes application:
+
+    kubectl create deployment hello-node --image=k8s.gcr.io/serve_hostname
+
+{{< / highlight >}}
+
+When that is done, the test can finally be executed successfully:
+
+{{< highlight bash >}}
+$ ci-operator --namespace bbguimaraes --config test.yaml --target test
+INFO[2022-03-10T14:41:33Z] unset version 0
+INFO[2022-03-10T14:41:33Z] Resolved source https://github.com/openshift/ci-tools to master@ef415037
+INFO[2022-03-10T14:41:35Z] Using namespace https://console.build01.ci.openshift.org/k8s/cluster/projects/bbguimaraes
+INFO[2022-03-10T14:41:35Z] Running [input:fedora], test
+INFO[2022-03-10T14:41:37Z] Tagging ci/fedora:latest into pipeline:fedora.
+INFO[2022-03-10T14:41:37Z] Executing test test
+INFO[2022-03-10T14:41:49Z] Ran for 16s
+{{< / highlight >}}
```

Comment:
NIT: I have a warning line there. You must have some permission that I do not have. 😄 

```console
# tried with build02
oc new-project hongkliu

ci-operator --namespace hongkliu --config test.yaml --target test
INFO[2023-02-22T11:19:42-05:00] unset version 0                              
INFO[2023-02-22T11:19:42-05:00] Resolved source https://github.com/openshift/ci-tools to master@ef415037 
INFO[2023-02-22T11:19:46-05:00] Using namespace https://console.build02.ci.openshift.org/k8s/cluster/projects/hongkliu 
INFO[2023-02-22T11:19:46-05:00] Running [input:fedora], test                 
WARN[2023-02-22T11:19:48-05:00] Could not edit namespace because you do not have permission to update the namespace.  error=<nil>
INFO[2023-02-22T11:19:49-05:00] Tagging ci/fedora:latest into pipeline:fedora. 
INFO[2023-02-22T11:19:49-05:00] Executing test test                          
INFO[2023-02-22T11:22:41-05:00] Ran for 2m59s
```

BTW, `2m59s` vs your `16s`. Gee~!

Reply:
That's probably either an image pull or due to the GCS credentials.  I was able to get a `24s` execution in `build02` after a couple of attempts.

This document is based on a somewhat idealized interaction with `ci-operator`.  I made several changes to it to get to this state, but some are still missing (the namespace warning — which I also get — and the GCS secret being two which I'm still working on).